### PR TITLE
controller/clusterdeployment: rate limit updates to clusterdeployment because of error

### PR DESCRIPTION
**Pull request by abhinavdahiya**
/assign @suhanime 
/cc @dgoodwin

### CODE REVIEW

**Comment by dgoodwin** [comment]
Looks good to me.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 1]
```diff
@@ -2013,3 +2015,43 @@ func LoadReleaseImageVerifier(config *rest.Config) (verify.Interface, error) {
...
+		}
+
+		if controllerutils.IsConditionInDesiredState(*oldcond) {
+			return true // newly Failed condition
+		}
+
+		if cond.Message != oldcond.Message ||
+			cond.Reason != oldcond.Reason {
+			return true // already failing but change in error reported
+		}
```
Comment (suhanime) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:2052]: We have something like this that we pass on as a criteria when we try to [update conditions](https://github.com/openshift/hive/blob/4376099acc9c78577b1609347655c98bc98def28/pkg/controller/utils/conditions.go#L28-L33)
This takes away the logic from the setting of conditions
Is there a way that we have the `UpdateConditionAlways`, `UpdateConditionNever`, `UpdateConditionIfReasonOrMessageChange` logic in only 1 place?

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 2]
```diff
@@ -2013,3 +2015,43 @@ func LoadReleaseImageVerifier(config *rest.Config) (verify.Interface, error) {
...
+	}
+
+	for _, cond := range new.Status.Conditions {
+		if controllerutils.IsConditionInDesiredState(cond) {
+			continue
+		}
+
+		oldcond := controllerutils.FindClusterDeploymentCondition(old.Status.Conditions, cond.Type)
+		if oldcond == nil {
+			return true // newly added condition in failed state
```
Comment (suhanime) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:2042]: failed state - or conditions were just initialized and set to unknown

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 3]
```diff
@@ -2013,3 +2015,43 @@ func LoadReleaseImageVerifier(config *rest.Config) (verify.Interface, error) {
...
+
+	old, ok := evt.ObjectOld.(*hivev1.ClusterDeployment)
+	if !ok {
+		return false
+	}
+
+	for _, cond := range new.Status.Conditions {
+		if controllerutils.IsConditionInDesiredState(cond) {
+			continue
+		}
```
Comment (suhanime) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:2038]: Wouldn't we want to update once if it was in undesired state before and now it is in desired state? We would want to give the signal that the error condition is cleared...

Comment (abhinavdahiya) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:2038]: the function decides if the the update event was due to error. it doesn't create signals for changes we must take care of.

in this scenario when we go from undesired to desired, the function returns false (unless there was another error change in another condition) and the update event is immediately put in the queue for reconcile.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 2 above)
Comment (abhinavdahiya) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:2042]: too specific for comment to be useful long term. unknown is undesired state for this context.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 1 above)
Comment (abhinavdahiya) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:2052]: > Is there a way that we have the UpdateConditionAlways, UpdateConditionNever, UpdateConditionIfReasonOrMessageChange logic in only 1 place?

that goal for update criteria is different from what we want to do here. updating conditions requires allowing caller to decide when to update whereas here we are calculating if the change in conditions was due to error. Using a function like UpdateConditionIfReasonOrMessageChange here creates more confusion than clarity.

**Pull request merged**